docs/rfcs/025-generation-numbers.md

```diff
@@ -0,0 +1,1119 @@
+  being certain that the previous pageserver has detached or stopped.
+- Server behavior: increment latest attach gen by 1 and return new attach_gen
+- Client behavior: the next attachment of the tenant should use the returned attach_gen. It is forbidden
+  to attach a tenant to a different pageserver without calling this API, unless the old pageserver has
+  return 200 to a detach request. In practice this API should be called between all attachment changes.
+
+For use in automated tests of the pageserver, a stub implementation would provide all three endpoints.
+
+### Timeline/Branch creation
+
+All of the previous arguments for safety have described operations within
+a timeline, where we may describe a sequence that includes updates to
+index_part.json, and where reads and writes are coming from a postgres
+endpoint (writes via the safekeeper).
+
+Creating or destroying timeline is a bit different, because writes
+are coming from the control plane.
+
+We must be safe against scenarios such as:
+
+- A tenant is attached to pageserver B while pageserver A is
+  in the middle of servicing an RPC from the control plane to
+  create or delete a tenant.
+- A pageserver A has been sent a timeline creation request
+  but becomes unresponsive. The tenant is attached to a
+  different pageserver B, and the timeline creation request
+  is sent there too.
+
+To properly complete a timeline create/delete request, we must
+be sure _after_ the pageserver write to remote storage, that its
+generation number is still up to date:
+
+- The pageserver can do this by calling into the control plane
+  before responding to a request.
+- Alternatively, the pageserver can include its generation in
+  the response, and let the control plane validate that this
+  generation is still current after receiving the response.
+
+If some very slow node tries to do a timeline creation _after_
+a more recent generation node has already created the timeline
+and written some data into it, that must not cause harm. This
+is provided in timeline creations by the way all the objects
+within the timeline's remote path include a generation suffix:
+a slow node in an old generation that attempts to "create" a timeline
+that already exists will just emit an index_part.json with
+an old generation suffix.
+
+Timeline IDs are never reused, so we don't have
+to worry about the case of create/delete/create cycles.
+
+**During timeline/tenant deletion, the control plane must not regard an operation
+as complete when it receives a `202 Accepted` response**, because the node
+that sent that response might become permanently unavailable. The control
+plane must wait for the deletion to be truly complete (e.g. by polling
+for 404 while the tenant is still attached to the same pageserver), and
+handle the case where the pageserver becomes unavailable, either by waiting
+for a replacement with the same node_id, or by re-attaching the tenant elsewhere.
```

Comment:
Attaching might not work anymore if the deletions have in fact succeeded.

We should just have a 'delete this tenant / timeline from S3 state' pageserver API that may or may not bypass the deletion queue.
And rely on scrubber for all the cases where that scheme can cause leakage.
Cases that I know will cause leakage:
* loss of in-memory state of deletion queue
* loss of node => loss of deletion queue unless we re-use node_id (we don't reuse node_id currently)
* stale attachment that does a write after we successfully did a one-off deletion pass

Reply:
Ah, a few paragraphs down, you elaborate on this.

Reply:
Yeah, the way we currently require attachment in order to do deletion is something we'll want to change.  It should make things quite a bit simpler.